File: agent-driven-architecture-main/examples/typescript/src/blocks/triage/tools.ts
// ── blocks/triage/tools — the Verb table ───────────────────────────────────
// One row per verb: name, model-facing description, input schema, the PURE run,
// the name→Command entry, and the reversibility classification. There is no
// default classification — you pick a sealed variant, which is how 14.3's
// default-deny becomes structural instead of remembered.
//
// The tool returns RAW INPUTS ONLY. It does not read state to compute
// `supersedes`, does not stamp an actor, and does not decide whether the
// transition is legal — the fold does all three (4.3's division of labour).
//
// The input schema for `ticket` is a plain string ON PURPOSE (6.10): the ticket
// set is OPEN at the boundary, and the ARM is what validates it against State.
//
// THE UPCASTER LIVES HERE, and not in contract.ts, because check C7 puts it
// here: a block mints its own ToolResults in this file and nowhere else, and
// `upcastSetPriority` produces one. The v1 SHAPE is declared next to the v2
// shape it evolved from (contract.ts); the LIFT is a production site.

import type { Verb } from "@adr/spine/pure/verb";
import { reversible } from "@adr/spine/pure/verb";
import type { InferOutput } from "valibot";
import { object, optional, picklist, string } from "valibot";
import type { SetPriorityCommand, SetPriorityResult, SetPriorityResultV1 } from "./contract";

const priority = picklist(["Low", "Normal", "High", "Urgent"]);

/** What a v1 record's missing `reason` becomes on the way into the fold (14.7).
 *  NOT `null`: `null` is v2's word for "the caller gave none", and a v1 record
 *  never had the field at all. An upcaster that erased that distinction would
 *  be inventing history rather than lifting it. */
export const PRE_V2_REASON = "not recorded (pre-v2 record)";

/** THE WORKED UPCASTER (14.7), v1 → v2 for this block's one payload.
 *
 *  Pure and total: every v1 result has exactly one v2 form. It never touches
 *  the record it came from — the caller gets a NEW value on the way into the
 *  fold, which is the difference between upcasting and the history rewrite
 *  14.1 forbids.
 *
 *  It re-mints `outcome` and `tool` rather than spreading `old`, which is not
 *  ceremony: the v1 payload's `outcome` is `"ok-v1"` — the conflict that makes
 *  a historical payload un-foldable — so the current shape has to be written
 *  out, and this function is the one place in the port that may write it. */
export function upcastSetPriority(old: SetPriorityResultV1): SetPriorityResult {
  return {
    outcome: "ok",
    tool: "setPriority",
    ticket: old.ticket,
    level: old.level,
    reason: PRE_V2_REASON,
  };
}

export function triageVerbs<S>(): readonly Verb<S>[] {
  return [
    reversible<
      S,
      { ticket: string; level: InferOutput<typeof priority>; reason?: string },
      SetPriorityResult,
      SetPriorityCommand
    >({
      name: "setPriority",
      describe: "Set a support ticket's priority (Low | Normal | High | Urgent).",
      // `reason` is the v2 field, and it is OPTIONAL at the schema too: an
      // adopter's existing callers keep working, which is the other half of
      // what 14.7's "optional field" buys.
      schema: object({ ticket: string(), level: priority, reason: optional(string()) }),
      // ── THE MODEL-FACING SURFACE, DECLARED IN THE BLOCK (SDK-1) ────────────
      // These cost NOTHING outside this folder. Before the Verb carried them,
      // no block could express any of it — the adapter was a generic converter
      // with nowhere to read a per-tool intent from, so the whole model-facing
      // half of the runtime's tool definition was unreachable by construction.
      //
      // `examples` earns its keep on this verb specifically: `level` is a closed
      // set and `reason` is optional, which a description can only gesture at.
      examples: [
        { ticket: "4118", level: "Urgent", reason: "customer reports funds taken twice" },
        { ticket: "4119", level: "Low" },
      ],
      // WHAT THE MODEL SEES, not what the timeline records. The committed result
      // is produced at the boundary from the raw input and is untouched by this
      // (C7); this only stops the reasoner re-reading a payload it already knows
      // it sent, which is the cheapest context there is to not spend.
      toModelOutput: (result) => `${result.ticket} → ${result.level}`,
      run: (input) => ({
        outcome: "ok",
        tool: "setPriority",
        ticket: input.ticket,
        level: input.level,
        reason: input.reason ?? null,
      }),
      // The Command does NOT mirror the new field, and the asymmetry is 14.7
      // doing its job: a Command is the SIGNED record of what a principal
      // authorized, and 14.1 forbids rewriting it — so the Command shape that
      // was signed is the shape that stays. What the fold consumes is the
      // RESULT, and that is what evolves.
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "setPriority",
        sig,
        id,
        ticket: result.ticket,
        level: result.level,
      }),
    }),
  ];
}

File: agent-driven-architecture-main/examples/kotlin/block/triage/src/main/kotlin/adr/blocks/triage/Tools.kt
// ── blocks/triage/tools — the Verb table ───────────────────────────────────
// One row per verb: name, description, input schema, PURE body, name→Command entry,
// reversibility. Registering a tool FORCES the reversibility decision; there is no
// default (14.3 default-deny, made structural).
//
// The tools are pure: they read Ctx and return a payload. They cannot name an
// Actor, an Authority or a Signature (gate check C4) — the stamp does not exist yet
// when they run.
//
// The input schema stays a plain string for `ticket`: the ticket set is OPEN at the
// boundary (6.10), and the ARM is what validates against state (12.4).
//
// The table and its decoder are members of a CONSTRUCTED type. `lens` is the one
// argument the split rule promotes to constructor state: it is fixed for a whole
// registration and shared by every row, exactly as the registry is shared by the
// boundary's two maps. Triage's single row does not happen to READ the lens — its arm
// validates against the slice instead — but the parameter is what keeps all six blocks
// registering identically at the root, and blocks/escalation/tools shows the row that
// genuinely needs it.
//
// `decodeSetPriority` is now PRIVATE TO THE TABLE THAT USES IT. `::decodeSetPriority`
// still reads the same at the call site; it is a bound reference to this instance's
// member rather than a pointer to a file-scope function nothing owns.
//
// THE UPCASTER LIVES HERE, and not in Contract.kt, because gate check C7 puts it
// here: a block mints its own ToolResults in this file and nowhere else, and
// `TriageUpcast.v1` produces one. The v1 SHAPE is declared next to the v2 shape it
// evolved from (Contract.kt); the LIFT is a production site.

package adr.blocks.triage

import adr.contract.TriageCommand
import adr.contract.TriageResult
import adr.contract.TriageResult.Priority
import adr.contract.TriageV1Result
import adr.spine.pure.Lens
import adr.spine.pure.RawInput
import adr.spine.pure.TicketId
import adr.spine.pure.ToolName
import adr.spine.pure.Verb

val SET_PRIORITY = ToolName("setPriority")

/**
 * What a v1 record's missing `reason` becomes on the way into the fold (14.7).
 *
 * NOT `null`: `null` is v2's word for "the caller gave none", and a v1 record never
 * had the field at all. An upcaster that erased that distinction would be inventing
 * history rather than lifting it.
 */
const val PRE_V2_REASON = "not recorded (pre-v2 record)"

internal data class SetPriorityInput(val ticket: TicketId, val level: Priority, val reason: String?)

/**
 * THE WORKED UPCASTER (14.7), v1 -> v2 for this block's one payload.
 *
 * Pure and total: every v1 result has exactly one v2 form. It holds nothing, so the
 * constructor is empty — the honest result of the split rule, as in TriageArm. It
 * never touches the record it came from; the caller gets a NEW value on the way into
 * the fold, which is the difference between upcasting and the history rewrite 14.1
 * forbids.
 */
class TriageUpcast {

    fun v1(old: TriageV1Result.SetPriority): TriageResult.SetPriority =
        TriageResult.SetPriority(old.tool, old.ticket, old.level, PRE_V2_REASON)
}

internal class TriageTools<S>(private val lens: Lens<S, TriageSlice>) {

    fun verbs(): List<Verb<S, *, *>> = listOf(
        Verb.Reversible(
            name = SET_PRIORITY,
            describe = "Set a support ticket's priority (Low | Normal | High | Urgent).",
            decode = ::decodeSetPriority,
            run = { input, _ ->
                TriageResult.SetPriority(SET_PRIORITY, input.ticket, input.level, input.reason)
            },
            // The Command does NOT mirror the new field, and the asymmetry is 14.7
            // doing its job: a Command is the SIGNED record of what a principal
            // authorized, and 14.1 forbids rewriting it — so the Command shape that was
            // signed is the shape that stays. What the fold consumes is the RESULT.
            sign = { r, sig, id -> TriageCommand.SetPriority(r.tool, sig, id, r.ticket, r.level) },
            narrow = { it as? TriageResult.SetPriority },
        ),
    )

    private fun decodeSetPriority(raw: RawInput): SetPriorityInput? {
        val ticket = raw.text("ticket") ?: return null
        val level = raw.text("level")?.let { name -> Priority.entries.firstOrNull { it.name == name } }
            ?: return null
        // `reason` is the v2 field and it is OPTIONAL at the schema too: an adopter's
        // existing callers keep working, which is the other half of what 14.7's
        // "optional field" buys. A missing field decodes to null, never to a failure.
        return SetPriorityInput(TicketId(ticket), level, raw.text("reason"))
    }
}
